File: Distinct_Subsequences.cpp
class Solution {
public:    
    /**
     * @param S, T: Two string.
     * @return: Count the number of distinct subsequences
     */
    int numDistinct(string &S, string &T) {
        // write your code here
        const int s_len = S.length();
        const int t_len = T.length();
        // vector<vector<int>> table(2, vector<int>(s_len + 1, 0));
        // table[0][0] = 1;
        // for (int i = 1; i < t_len + 1; ++i){
        //     table[i % 2][0] = 0;
        //     for (int j = 1; j < s_len + 1; ++j){
        //         table[i % 2][j] = table[i % 2][j - 1];
        //         if (T[i - 1] == S[j - 1]){
        //             table[i % 2][j] += table[(i - 1)% 2][j - 1] == 0 && i == 1 ? 1 : table[(i - 1)% 2][j - 1];
        //         }

        //     }
        // }
        // return table[t_len % 2][s_len];
        //other solution, the other way
        vector<vector<int>> table(2, vector<int>(t_len + 1, 0));
        table[0][0] = 1;
        for (int i = 1; i < s_len + 1; ++i){
            table[i % 2][0] = 1;// table[i % 2][0] = table[(i - 1) % 2][0];
            for (int j = 1; j < t_len + 1; ++j){
                // table[i % 2][j] += table[(i - 1)][j];
                table[i %2][j] = S[i - 1] == T[j - 1] ? table[(i - 1) % 2][j] + table[(i - 1) % 2][j - 1] : table[(i - 1) % 2][j];
            }
        }
        return table[s_len % 2][t_len];
    }
};

File: Distinct Subsequences.py
class Solution: 
    # @param S, T: Two string.
    # @return: Count the number of distinct subsequences
    def numDistinct(self, S, T):
        # write your code here
        
        m, n = len(S), len(T)
        
        dp = [[0] * (n + 1) for __ in range(m + 1)]
        dp[0][0] = 1  # error
        for i in range(1, m + 1):
            dp[i][0] = 1 # error
                          # error
            for j in range(1, n + 1):
                dp[i][j] = dp[i - 1][j]
                if S[i - 1] == T[j - 1]:
                    dp[i][j] += dp[i - 1][j - 1]
        return dp[m][n]
                